pyiron_atomistics/sphinx/output_parser.py

```diff
@@ -386,3 +388,108 @@ def get_convergence(self):
     def get_fermi(self):
         pattern = r"Fermi energy:\s+(\d+\.\d+)\s+eV"
         return np.array(re.findall(pattern, self.log_main)).astype(float)
+
+
+class SphinxWavesParser:
+    """ Class to read SPHInX waves.sxb files (HDF5 format)
+    
+        Initialize with waves.sxb filename, or use load ()
+    
+    """
+    
+    def __init__(self, file_name="waves.sxb", cwd="."):
+        """
+        Args:
+            file_name (str): file name
+            cwd (str): directory path
+        """
+        self._eps = None
+        if Path(file_name).is_absolute():
+            self.wfile = h5py.File(Path(file_name))
+        else:
+            path = Path(cwd) / Path(file_name)
+            self.wfile = h5py.File(path)
+        
+
+    @property
+    def _n_gk(self):
+        return self.wfile['meshDim'][0]  
+    
+    @property
+    def _fft_idx(self):
+        fft_idx=[]
+        off=0
+        for ngk in self._n_gk:
+            fft_idx.append (self.wfile['fftIdx'][off:off+ngk])
+            off += ngk
+        return fft_idx
+    
+    @property
+    def mesh(self):
+        return self.wfile['meshDim'][:]
+
+    @property
+    def Nx(self):
+        return self.wfile['meshDim'][0]  
+
+    @property
+    def Ny(self):
+        return self.wfile['meshDim'][1]  
+
+    @property
+    def Nz(self):
+        return self.wfile['meshDim'][2]        
+        
+    @property
+    def n_states(self):
+        return self.wfile["nPerK"][0]
+
+    @property
+    def n_spin(self):
+        return self.wfile['nSpin'].shape[0]
+
+    @property
+    def k_weights(self):
+        return self.wfile['kWeights'][:]
+
+    @property
+    def k_vec(self):
+        return self.wfile['kVec'][:]
+
+    @property
+    def eps(self):
+        """All eigenvalues (in Hartree) as (nk,n_states) block"""
+        if (self._eps is None):
+            self._eps = self.wfile['eps'][:].reshape (-1,self.n_spin,self.n_states)
+        return self._eps.T #change
+
+    # Define as separate method and speed it up with numba
+    @staticmethod
+    @numba.jit
+    def _fillin(res,psire,psiim,fft_idx):
+        """Distributes condensed psi (real, imag) on full FFT mesh"""
+        rflat=res.flat
+        for ig in range(fft_idx.shape[0]):
+            rflat[fft_idx[ig]] = complex(psire[ig], psiim[ig])
+
+    def get_psi_rec(self,i, ispin, ik):
```

Comment:
I'm generally very uncomfortable with the way this function is written. Is there maybe a possibility to slice `wfile` at the beginning to avoid the offsets? @freyso

Reply:
as  far as I understand hdf5 python interface, wfile["varname"] does not read anything, it returns an abstract handle for the data inside the file. Only when I index the handle, the data is read into a numpy array.